include/engine/guidance/api_response_generator.hpp

```diff
@@ -0,0 +1,286 @@
+    typedef detail::Segment Segment;
+    typedef ExtractRouteNames<DataFacade, Segment> RouteNameExtractor;
+
+    ApiResponseGenerator(DataFacade *facade);
+
+    // This runs a full annotation, according to config.
+    // The output is tailored to the viaroute plugin.
+    void DescribeRoute(const DescriptorConfig &config,
+                       const InternalRouteResult &raw_route,
+                       osrm::json::Object &json_result);
+
+    // The following functions allow access to the different parts of the Describe Route
+    // functionality.
+    // For own responses, they can be used to generate only subsets of the information.
+    // In the normal situation, Describe Route is the desired usecase.
+
+    // generate an overview of a raw route
+    osrm::json::Object SummarizeRoute(const InternalRouteResult &raw_route,
+                                      const SegmentListT &segment_list) const;
+
+    // create an array containing all via-points/-indices used in the query
+    osrm::json::Array ListViaPoints(const InternalRouteResult &raw_route) const;
+    osrm::json::Array ListViaIndices(const SegmentListT &segment_list) const;
+
+    osrm::json::Value GetGeometry(const bool return_encoded, const SegmentListT &segments) const;
+
+    // TODO this dedicated creation seems unnecessary? Only used for route names
+    std::vector<Segment> BuildRouteSegments(const SegmentListT &segment_list) const;
+
+    // adds checksum and locations
+    osrm::json::Object BuildHintData(const InternalRouteResult &raw_route) const;
+
+  private:
+    // data access to translate ids back into names
+    DataFacade *facade;
+};
+
+template <typename DataFacadeT>
+ApiResponseGenerator<DataFacadeT>::ApiResponseGenerator(DataFacadeT *facade_)
+    : facade(facade_)
+{
+}
+
+template <typename DataFacadeT>
+void ApiResponseGenerator<DataFacadeT>::DescribeRoute(const DescriptorConfig &config,
+                                                      const InternalRouteResult &raw_route,
+                                                      osrm::json::Object &json_result)
+{
+    if( not raw_route.is_valid() ){
+      return;
+    }
+    const bool ALLOW_SIMPLIFICATION = true;
+    const bool EXTRACT_ROUTE = false;
+    const bool EXTRACT_ALTERNATIVE = true;
+    SegmentListT segment_list(raw_route, EXTRACT_ROUTE, config.zoom_level, ALLOW_SIMPLIFICATION,
+                              facade);
+    json_result.values["route_summary"] = SummarizeRoute(raw_route, segment_list);
+    json_result.values["via_points"] = ListViaPoints(raw_route);
+    json_result.values["via_indices"] = ListViaIndices(segment_list);
+
+    if (config.geometry)
+    {
+        json_result.values["route_geometry"] = GetGeometry(config.encode_geometry, segment_list);
+    }
+
+    if (config.instructions)
+    {
+        json_result.values["route_instructions"] =
+            annotators::AnnotateRoute(segment_list.Get(), facade);
```

Comment:
This is the call site that would have to decide on the policy to use if we want to make the annotation generic